developers-map.md

```diff
@@ -0,0 +1,171 @@
+    - passes events to plugin manager
+    - interface: [./pkg/secretless/plugin/v1/event_notifier.go](./pkg/secretless/plugin/v1/event_notifier.go)
+    - implementation: [./internal/pkg/plugin/manager.go](./internal/pkg/plugin/manager.go)
+  + [Listener](#listener)
+    - listens and accepts client connections and passes them to a handler
+    - interface: [./pkg/secretless/plugin/v1/listener.go](./pkg/secretless/plugin/v1/listener.go)
+    - pg implementation: [./internal/app/secretless/listeners/pg/listener.go](./internal/app/secretless/listeners/pg/listener.go)
+  + [Handler](#handler)
+    - receives client connection, connect it to a backend and streams connection.
+    - interface: [./pkg/secretless/plugin/v1/handler.go](./pkg/secretless/plugin/v1/handler.go)
+    - pg implementation: [./internal/app/secretless/handlers/pg/handler.go](./internal/app/secretless/handlers/pg/handler.go)
+  + Provider
+    - used to obtain values from a secret vault backend
+    - interface: [./pkg/secretless/plugin/v1/provider.go](./pkg/secretless/plugin/v1/provider.go)
+    - env implementation: [./internal/app/secretless/providers/env/provider.go](./internal/app/secretless/providers/env/provider.go)
+  + [Resolver](#resolver)
+    - manages Providers and provides convenient interface to obtain multiple values from multiple secret vault backends
+    - interface: [./pkg/secretless/plugin/v1/resolver.go](./pkg/secretless/plugin/v1/resolver.go)
+    - implementation: [./internal/pkg/plugin/resolver.go](./internal/pkg/plugin/resolver.go)
+ 
++ internal implementations of most plugin interfaces
+  + located at ./internal/app/secretless
+  + most plugins are exposed as factories, where the key is the identifier and the value is an interface pointer e.g. map[string]func(plugin_v1.HandlerOptions) plugin_v1.Handler
+
+### EventNotifier
+interface: [./pkg/secretless/plugin/v1/event_notifier.go](./pkg/secretless/plugin/v1/event_notifier.go)
+
+implementation: [./internal/pkg/plugin/manager.go](./internal/pkg/plugin/manager.go)
+
++ mechanism for bubbling up events such as new connection/handler/listener
++ plugin manager is the only implementation
++ threaded from plugin manager to relevant children
++ used by
+  + Listener
+  + Handler
+  + Resolver
+  + Proxy
+
+### Proxy 
+secretless.Proxy
+
+[./internal/app/secretless/proxy.go](./internal/app/secretless/proxy.go)
+
++ takes
+  + EventNotifier
+  + [Resolver](#resolver)
+  + RunListenerFunc - ListenerFactory
+    + takes 
+      + listener id
+      + ListenerOptions
+    + returns Listener interface pointer
+  + RunHandlerFunc - HandlerFactory, passed to listener
+    + NOTE: HandlerFactories run Handlers before returning them.
+    + takes 
+      + handler id
+      + HandlerOptions
+    + returns Listener interface pointer
++ manages Listener lifecycles
+  + in #Listen, for each listener 
+    + creates net.Listener based on config.Config
+    + creates Listener interface pointer using RunListenerFunc
```

Comment:
consider for later doing this with some generator or the docs will get out of date quick